docs/architecture/how/receipt-congestion.md

```diff
@@ -0,0 +1,211 @@
+
+## Idea 2: Limit transactions when we use too much memory
+
+What if we have no limit at the source until we notice we are above the memory
+threshold we are comfortable with? Then we can reduce the source capacity in
+steps, potentially down to 0, until buffers are getting emptier and we use less
+memory again.
+
+If we do that, we can decide between either applying a global limit on all
+sources (allow only `1000 Tgas / NUM_SHARDS` new transactions on all shards like
+in idea 1) or applying the limit only to transactions that go to the shard with
+the congestion problem.
+
+The first choice is certainly safe. But it means that a single congested shard
+leads to all shards slowing down, even if they could keep working faster without
+ever sending receipts to the congested shard. This is a hit to utilization we
+want to avoid. So let's try the second way.
+
+In that case we filter transactions by receiver and keep accepting transactions
+that go to non-congested shards. This would work fine, if all transactions would
+only have depth 1.
+
+But receipts produced by an accepted transaction can produce more receipts to
+any other shard. Therefore, we might end up accepting more inflow that
+indirectly requires bandwidth on the congested shard.
+
+![graph](../../images/congestion/receipt_flow_example_2.svg)
+
+Crucially, when accepting a transaction, we don't know ahead of time which
+shards will be affected by the full directed graph of receipts in a transaction.
+We only know the first step. For multi-jop transactions, there is no easy way out.
+
+But it is worth mentioning, that in practice the single-hop function call is the
+most common case. And this case can be handled nicely by rejecting incoming
+transactions to congested shards.
+
+## Idea 3: Apply backpressure to stop all flows to a congested shard
+
+On top of stopping transactions to congested shards, we can also stop receipts if they have a congested shard as the receiver.
+We simply put them in a buffer of the sending shard and keep them there until
+the congested shard has space again for the receipts.
+
+![graph](../../images/congestion/receipt_flow_example_3.svg)
+
+The problem with this idea is that it leads to deadlocks where all receipts in
+the system are waiting in outgoing buffers but cannot make progress because the
+receiving shard already has too high memory usage.
+
+![graph](../../images/congestion/receipt_flow_example_3_1.svg)
+
+## Idea 4: Keep minimum incoming queue length to avoid deadlocks
+
+This is the final idea we need. To avoid deadlocks, we ensure that we can always
+send receipts to a shard that has not enough work in the delayed receipts queue
+already.
+
+Basically, the backpressure limits from idea 3 are only applied to incoming
+receipts but not for the total size. This guarantees that in congested the
+scenario that previously caused a deadlock, we always have something in the
+incoming queue to work on, otherwise there wouldn't be backpressure at all.
```

Comment:
```suggestion
Basically, the backpressure limits from idea 3 are only applied to incoming
receipts but not for the total size. This guarantees that in the congested
scenario that previously caused a deadlock, we always have something in the
incoming queue to work on, otherwise there wouldn't be backpressure at all.
```